cpp/test/sg/svc_test.cu

```diff
@@ -0,0 +1,891 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuda_utils.h>
+#include <gtest/gtest.h>
+#include <test_utils.h>
+#include <cub/cub.cuh>
+#include <iostream>
+#include "common/cumlHandle.hpp"
+#include "gram/grammatrix.h"
+#include "gram/kernelmatrices.h"
+#include "svm/smosolver.h"
+#include "svm/svc.h"
+#include "svm/workingset.h"
+#include "test_utils.h"
+
+namespace ML {
+namespace SVM {
+using namespace MLCommon;
+using namespace GramMatrix;
+
+class WorkingSetTest : public ::testing::Test {
```

Comment:
These tests are well done. We should also be including testing double precision for every applicable algorithm. Can you provide a test for the lower bound of single precision in the solver? (eg quantify the truncation bounds / underflow potential, etc...)

We have a new digits dataset in `src/datasets/digits.h` and are also working to increase our code coverage of tests using different datasets. We also have a `make_blobs` in the C++ layer, which should allow you to test using blobs of known covariance.

Reply:
I do not see `src/datasets/digits.h`, but I used `make_blobs` to test on datasets with different shapes. I see that the larger blob examples  are sensitive to the precision, I think it should be related to the fact that the input is not scaled. I still have to confirm that this is the expected behavior.